maintenance/program/src/instruction.rs

```diff
@@ -0,0 +1,313 @@
+//! Program instructions
+
+use borsh::{BorshDeserialize, BorshSchema, BorshSerialize};
+use solana_program::{
+    bpf_loader_upgradeable,
+    instruction::{AccountMeta, Instruction},
+    hash::Hash,
+    pubkey::Pubkey,
+    system_program,
+    sysvar,
+};
+
+use crate::{
+    processor::get_maintenance_record_seeds,
+};
+
+/// Instructions supported by the VoterWeight addin program
+/// This program is a mock program used by spl-governance for testing and not real addin
+#[derive(Clone, Debug, PartialEq, BorshDeserialize, BorshSerialize, BorshSchema)]
+#[allow(clippy::large_enum_variant)]
+pub enum MaintenanceInstruction {
+    /// Creates MaintenanceRecord owned by the program
+    ///
+    /// 0. `[]` Maintained program account
+    /// 1. `[writable]` MaintenanceRecord
+    /// 2. `[signer]` Authority
+    /// 3. `[signer]` Payer
+    /// 4. `[]` System
+    CreateMaintenance { },
+
+    /// Sets Delegates into MaintenanceRecord
+    ///
+    /// 0. `[writable]` MaintenanceRecord
+    /// 1. `[signer]` Authority
+    /// 2. `[signer]` Payer
+    SetDelegates {
+        #[allow(dead_code)]
+        delegate: Vec<Pubkey>,
+    },
+
+    /// Sets Code Hashes into MaintenanceRecord
+    ///
+    /// 0. `[writable]` MaintenanceRecord
+    /// 1. `[signer]` Authority
+    /// 2. `[signer]` Payer
+    SetCodeHashes {
+        #[allow(dead_code)]
+        hashes: Vec<Hash>,
+    },
+
+    /// Upgrades the Maintained program from buffer
+    ///
+    /// 0. `[]` Bpf Loader Upgradeable Program Id
+    /// 1. `[]` Sysvar Rent Program Id
+    /// 2. `[]` Sysvar Clock Program Id
+    /// 3. `[writable]` Maintained program account
+    /// 4. `[writable]` Maintained program data account
+    /// 5. `[writable]` Upgrade buffer account
+    /// 6. `[]` MaintenanceRecord
+    /// 7. `[signer]` Authority
+    /// 8. `[signer]` Payer
+    Upgrade { },
+
+    /// Revokes Authority from the program
+    ///
+    /// 0. `[]` Bpf Loader Upgradeable Program Id
+    /// 1. `[]` Maintained program account
+    /// 2. `[writable]` Maintained program data account
+    /// 3. `[]` MaintenanceRecord
+    /// 4. `[signer]` Authority
+    /// 5. `[]` New Authority
+    /// 6. `[signer]` Payer
+    SetAuthority { },
+
+    /// Closes MaintenanceRecord owned by the program
+    ///
+    /// 0. `[writable]` MaintenanceRecord
+    /// 1. `[writable]` Spill destination
+    /// 2. `[]` Maintained program data account
```

Comment:
You should check the authority before closing the `MaintenanceRecord`!

Reply:
Fixed